certora/specs/GeneralAdapter1Reverts.spec

```
@@ -0,0 +1,135 @@
+    function _.deposit(uint256, address) external => DISPATCHER(true);
+    function _.withdraw(uint256, address, address) external => DISPATCHER(true);
+    function _.redeem(uint256, address, address) external => DISPATCHER(true);
+    function _.approve(address, uint256)  external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function AllowanceTransfer.allowance(address, address, address) external returns (uint160, uint48, uint48) envfree;
+    function Bundler3.initiator() external returns address envfree;
+}
+
+// Permit2Lib.PERMIT2 concrete address
+definition PERMIT2() returns address = 0x000000000022D473030F116dDEE9F6B43aC78BA3;
+
+// Check that the state didn't change upon a native transfer with 0 ETH using the adapter.
+rule nativeTransferChange(env e, address receiver, uint256 amount) {
+    storage storageBefore = lastStorage;
+    uint256 senderBalanceBefore = nativeBalances[currentContract];
+
+    nativeTransfer(e, receiver, amount);
+
+    // Equivalence is rewrittent to avoid issue with quantifiers and polarity.
+    assert (amount == max_uint256 && senderBalanceBefore == 0 => storageBefore == lastStorage) &&
+           storageBefore == lastStorage => amount == max_uint256 && senderBalanceBefore == 0;
+}
+
+// Check that the state didn't change upon unwrapping 0 ETH using the adapter.
+rule erc20TransferChange(env e, address token, address receiver, uint256 amount) {
+    uint256 senderBalanceBefore = token.balanceOf(e, currentContract);
+    storage storageBefore = lastStorage;
+
+    erc20Transfer(e, token, receiver, amount);
+
+    // Equivalence is rewrittent to avoid issue with quantifiers and polarity.
+    assert (amount == 0 || (senderBalanceBefore == 0 && amount == max_uint256) => storageBefore == lastStorage) &&
+           (storageBefore == lastStorage => amount == 0 || (senderBalanceBefore == 0 && amount == max_uint256));
+}
+
+// Check that state didn't change upon an ERC20 self-transfer using the adapter.
+rule erc20TransferFromChange(env e, address token, address receiver, uint256 amount) {
+
+    // Safe require as the initiator can't be the adatper.
+    require Bundler3.initiator() != currentContract;
+
+    uint256 senderBalanceBefore = token.balanceOf(e, currentContract);
+    storage storageBefore = lastStorage;
+
+    erc20TransferFrom(e, token, receiver, amount);
+
+    // Equivalence is rewrittent to avoid issue with quantifiers and polarity.
+    assert (receiver == Bundler3.initiator() && token.allowance(e, Bundler3.initiator(), currentContract) == max_uint256 => storageBefore == lastStorage) &&
+           (storageBefore == lastStorage => receiver == Bundler3.initiator() && token.allowance(e, Bundler3.initiator(), currentContract) == max_uint256);
```

(optional) could apply to other places where equivalence is rewritten as well
```suggestion
    bool noChangeExpectedCondition = receiver == Bundler3.initiator() && token.allowance(e, Bundler3.initiator(), currentContract) == max_uint256 ;
    assert (noChangeExpectedCondition => storageBefore == lastStorage) &&
           (storageBefore == lastStorage => noChangeExpectedCondition);
```

---

Great idea, done in https://github.com/morpho-org/bundler3/pull/229/commits/9248da91e0789074a349baf82cd83aa5ee13ce34.